shared/src/main/java/com/couchbase/lite/AbstractReplicator.java

```diff
@@ -371,6 +433,44 @@ public void removeChangeListener(ListenerToken token) {
         }
     }
 
+    /**
+     * Set the given DocumentReplicatedListener to the this replicator.
+     *
+     * @param listener
+     * @return ListenerToken A token to remove the handler later
+     */
+    public ListenerToken addReplicationListener(DocumentReplicatedListener listener) {
+        return addReplicationListener(null, listener);
+    }
+
+    /**
+     * Set the given DocumentReplicatedListener to the this replicator.
+     *
+     * @param listener
+     */
+    public ListenerToken addReplicationListener(Executor executor, DocumentReplicatedListener listener) {
+        synchronized (lock) {
+            if (listener == null)
+                throw new IllegalArgumentException();
+            this.config.setProgressLevel(ReplicatorConfiguration.ReplicatorProgressLevel.PER_DOCUMENT);
+            DocumentReplicatedListenerToken token = new DocumentReplicatedListenerToken(executor, listener);
+            docEndedListenerTokens.add(token);
+            return token;
+        }
+    }
+
+    /**
+     * Remove the given DocumentReplicatedListener from the this replicator.
+     */
+    public void removeReplicationListener(ListenerToken token) {
+        synchronized (lock) {
+            if (token == null || !(token instanceof DocumentReplicatedListenerToken))
+                throw new IllegalArgumentException();
+            this.config.setProgressLevel(ReplicatorConfiguration.ReplicatorProgressLevel.OVERALL);
```

Comment:
This logic is also not correct because there could be multiple `DocumentReplicationListener` added so need to set to `OVERALL` only when all of the `DocumentReplicationListener` objects added were removed.